src/MySqlConnector/MySqlClient/MySqlDataReader.cs

```diff
@@ -781,12 +830,32 @@ private enum State
 			NoMoreData,
 		}
 
+	    public async Task WaitDoneReadingAsync(CancellationToken cancellationToken=default(CancellationToken))
+	    {
+	        cancellationToken.ThrowIfCancellationRequested();
+	        await m_done_reading.WaitAsync(cancellationToken).ConfigureAwait(false);
+	        m_done_reading.Release();
+	    }
+
 		static readonly Task<bool> s_falseTask = Task.FromResult(false);
 		static readonly Task<bool> s_trueTask = Task.FromResult(true);
 
 		MySqlCommand m_command;
 		MySqlSession m_session;
-		State m_state;
+	    SemaphoreSlim m_done_reading = new SemaphoreSlim(1);
```

Comment:
If this semaphore starts with an initial count of 1, then the first call to `WaitAsync` will enter the semaphore immediately; thus this won't actually enforce mutual exclusion.

I'm also not sure that this would correctly handle the case of multiple threads calling `ExecuteAsync` and all waiting on the semaphore.